### PR TITLE
Routing topic 3.0 updates

### LINKED ISSUE #9552: Routing topic: Remove .NET version language and use whole-topic versioning
The Routing topic updates on #9527 explain new-for-2.2 endpoint-based routing in the context of legacy IRouter-based routing and refer explicitly to the 2.2 version release. Remove the versioning language (e.g., "new in").

Refactor for whole-topic versioning.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 32374077-e02e-d6a6-1cea-05e58447b7cd
* Version Independent ID: 86aeab13-1c25-9fc1-4e56-0390cb40c242
* Content: [Routing in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-2.1)
* Content Source: [aspnetcore/fundamentals/routing.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/fundamentals/routing.md)
* Product: **aspnet-core**
* GitHub Login: @ardalis
* Microsoft Alias: **riande**

### LINKED ISSUE #14291: 3.0 update routing with new Endpoint routing
Add Endpoint routing updates

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 32374077-e02e-d6a6-1cea-05e58447b7cd
* Version Independent ID: 86aeab13-1c25-9fc1-4e56-0390cb40c242
* Content: [Routing in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-3.0)
* Content Source: [aspnetcore/fundamentals/routing.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/fundamentals/routing.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #9552
Fixes #14291

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-2.1&branch=pr-en-us-14334)

* Should we work on #13245 with this update, or leave it for another time (or place given that it really is in an SPA topic)? I'm very 🏃🏃😅 busy. If it can wait, let's leave that open and we'll get to it later.
* I didn't see Announcement issues. Anything *new* for 3.0 for routing?
* New 3.0 sample app.
* Whole-topic versioning.
* In 3.0 content, I kept the *Differences from earlier versions of routing* section but removed some other 'this is new' language elsewhere.

WRT **_new for 3.0_**, I was only able to quickly piece together this for the release notes topic from the blog post (which is probably not exactly right) ...

> Endpoint Routing, which allows frameworks (for example, MVC) to work well with middleware, is enhanced:
>
> * The order of middleware and endpoints is configurable in the request processing pipeline of `Startup.Configure`.
> * Endpoints and middleware compose well with other ASP.NET Core-based technologies, such as Health Checks.
> * Endpoints can implement a policy, such as CORS or authorization, in both middleware and MVC.
> * Filters and attributes can be placed on methods in controllers.

Give me a blurb or troubleshoot that, and I'll get it in here.

On the *implement a policy* bit, I can cross-link. I think that work is done/shown in auth and Health Checks topics. Not sure about the *filters and attributes on methods* tho ... I'll need to look :eyes:.

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@rynowak please review or suggest a reviewer 
@mkArtakMSFT

File: aspnetcore/fundamentals/routing.md [hunk 1]
```diff
@@ -774,3 +585,1007 @@ Link generation only generates a link for this route when the matching values fo
...
 <!-- While that code is no longer used by ASP.NET Core for complex segment matching, it provides a good match to the current algorithm. The [current code](https://github.com/aspnet/AspNetCore/blob/91514c9af7e0f4c44029b51f05a01c6fe4c96e4c/src/Http/Routing/src/Matching/DfaMatcherBuilder.cs#L227-L244) is too abstracted from matching to be useful for understanding complex segment matching.
 -->
+
+::: moniker-end
+
+::: moniker range="= aspnetcore-2.2"
+
+Routing is responsible for mapping request URIs to endpoint selectors and dispatching incoming requests to endpoints. Routes are defined in the app and configured when the app starts. A route can optionally extract values from the URL contained in the request, and these values can then be used for request processing. Using route information from the app, routing is also able to generate URLs that map to endpoint selectors.
```
Comment (rynowak) [inline aspnetcore/fundamentals/routing.md]: I'd suggest dropping the term *endpoint selector* as it's used here and call refer to it as an *endpoint*.

`IEndpointSelector` is a very specific piece of extensibility that most users won't need, so it shouldn't show up here as a concept.

File: aspnetcore/fundamentals/routing.md [hunk 2]
```diff
@@ -2,65 +2,20 @@
...
 ms.custom: mvc
-ms.date: 06/17/2019
+ms.date: 09/16/2019
 uid: fundamentals/routing
 ---
 # Routing in ASP.NET Core
 
 By [Ryan Nowak](https://github.com/rynowak), [Steve Smith](https://ardalis.com/), and [Rick Anderson](https://twitter.com/RickAndMSFT)
 
-::: moniker range="<= aspnetcore-1.1"
```
Comment (rynowak) [inline aspnetcore/fundamentals/routing.md]: What's the gist of the moving around of content? It's hard for me to get a sense of what this doc is going to look like when someone reads the 3.0 version.

Comment (rynowak) [reply aspnetcore/fundamentals/routing.md]: It would be great if someone could give me an idea of what parts of this are new and need review.

**Comment by Rick-Anderson** [comment]
> It's hard for me to get a sense of what this doc is going to look like when someone reads the 3.0 version.

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-2.1&branch=pr-en-us-14334)

**Comment by guardrex** [comment]
I think that should be ...

https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-3.0&branch=pr-en-us-14334

**Comment by guardrex** [comment]
@Rick-Anderson Ignore the build warning ... that was fixed ages ago on another PR.

**Pull request merged**